### ISSUE TITLE
XPRESS interface: SetSolverSpecificParameters misbehaves for some parameters

### ISSUE THREAD

**Issue by pet-mit** [rte-france/or-tools#29]
**What version of OR-Tools and what language are you using?**
Version: v9.2 RTE1.1
Language:  Python

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
XPRESS

**What operating system (Linux, Windows, ...) and version?**
Windows

**What did you do?**
called solver.SetSolverSpecificParameters("PERTURB 0.0001 SBEFFORT 0.25")

**What did you expect to see**
In the XPRESS logs, I expected to see the following control parameters be set:
PERTURB = 0.0001
SBEFFORT = 0.25

**What did you see instead?**
Instead, they seem to have been rounded:
PERTURB = 0
SBEFFORT = 0

**Additional context**
*NB:* I tried calling solver.SetSolverSpecificParameters("PERTURB 1e-4 SBEFFORT 25e-2"), and it worked properly:
PERTURB = 0.0001
SBEFFORT = 0.25

**Comment by flomnes** [comment]
@pet-mit Your problem seems to be related to formatted input. XPRESS understands 0.0001 as 0, but 1e-4 as 1e-4.

A possible fix could be to format numbers properly (scientific notation) before passing them to XPRESS.

In the meantime, to avoid running into this issue, you use the right formatting from the start.

**Comment by flomnes** [comment]
Couldn't reproduce in the C++ unit tests defined in **ortools/linear_solver/unittests/xpress_interface.cc**.

The issue may be related to type conversions to/from Python.

**Comment by flomnes** [comment]
Couldn't reproduce.

```py
# Création du problème, choix du solveur
from ortools.linear_solver import pywraplp as pylp
solver=pylp.Solver.CreateSolver("xpress_lp")
solver.SetSolverSpecificParametersAsString("PERTURB 0.0001 SBEFFORT 25e-2")
#I0715 13:20:07.009526 95350 xpress_interface.cc:1891] Setting parameter SBEFFORT to value 0.0001
#I0715 13:20:07.009604 95350 xpress_interface.cc:1891] Setting parameter SBEFFORT to value 25e-2
```

**Comment by pet-mit** [comment]
thanks fro your feedback
so if i understand correctly, this issue is not solved by your pull request. i will have to use the workaround

**Comment by flomnes** [comment]
@pet-mit I think I found what's causing the error : Solve
```cpp
MPSolver::ResultStatus XpressInterface::Solve(MPSolverParameters const& param) {
  // ...
  // Set parameters.
  // NOTE: We must invoke SetSolverSpecificParametersAsString() _first_.
  //       Its current implementation invokes ReadParameterFile() which in
  //       turn invokes XPRSreadcopyparam(). The latter will _overwrite_
  //       all current parameter settings in the environment.
  solver_->SetSolverSpecificParametersAsString(
      solver_->solver_specific_parameter_string_);
  SetParameters(param);
  // ...
}
```

**Comment by flomnes** [comment]
So, it seems that the log was incorrect. @sgatto fixed it here 27f1956d98b620c4b3655c619fb171d6437f0e3a. This patch was integrated into v9.2-rte1.1 (your version). This is the bug I ran into, not yours @pet-mit, my bad.

**Comment by flomnes** [comment]
See [this page](https://en.cppreference.com/w/cpp/string/byte/strtof).
